React edit (#14019)
PrefectHQ/prefect: docs/3.0rc/react/understand-events/automations-triggers.mdx

---
title: Use automations for dynamic responses
description: Automations provide a flexible and powerful framework for automatically taking action in response to events. 
---

Automations are a Prefect Cloud feature that enable you to configure [actions](#actions) that execute automatically based on 
[trigger](#triggers) conditions.

Potential triggers include the occurrence of events from changes in a flow run's state—or the absence of such events. 
You can even define your own custom trigger to fire based on an [event](/3.0rc/react/understand-events/) created from a 
webhook or a custom event defined in Python code.

> Potential actions include starting flow runs, pausing schedules, and sending custom notifications.

Reviewer: ```suggestion
Actions include starting flow runs, pausing schedules, and sending custom notifications.
```

> Potential actions include kicking on flow runs, pausing schedules, and sending custom notifications.

Reviewer: ```suggestion
Potential actions include starting flow runs, pausing schedules, and sending custom notifications.
```

    "match": {
      "prefect.resource.id": "prefect.flow-run.*"
    },
    "expect": ["prefect.flow-run.Completed"]
  }'

# Pass a trigger using a JSON/YAML file
prefect deploy -n test-deployment --trigger triggers.yaml
prefect deploy -n test-deployment --trigger my_stuff/triggers.json
```

For example, a `triggers.yaml` file could have many triggers defined:

```yaml
triggers:
  - enabled: true
    match:
      prefect.resource.id: my.external.resource
    expect:
      - external.resource.pinged
    parameters:
      param_1: "{{ event }}"
  - enabled: true
    match:
      prefect.resource.id: my.other.external.resource
    expect:
      - some.other.event
    parameters:
      param_1: "{{ event }}"
```

Both of the above triggers would be attached to `test-deployment` after running `prefect deploy`.

<Warning>
**Triggers passed to `prefect deploy` will override any triggers defined in `prefect.yaml`**

While you can define triggers in `prefect.yaml` for a given deployment, triggers passed to `prefect deploy` 
take precedence over those defined in `prefect.yaml`.
</Warning>

Note that deployment triggers contribute to the total number of automations in your workspace.

> ## Automation notifications

Reviewer: ```suggestion
## Sending notifications with automations
```

> Notifications enable you to set up automation actions that send a message.

Reviewer: ```suggestion
```

> Automation notifications support sending notifications through any predefined block that is capable of and configured 

Reviewer: ```suggestion
Automations support sending notifications through any predefined block that is capable of and configured 
```

> to send a message. That includes, for example:

Reviewer: ```suggestion
to send a message, including:
```

![Viewing automations for a workspace in Prefect Cloud.](/3.0rc/img/ui/automations.png)

Select the toggle next to an automation to pause execution of the automation.

The button next to the toggle provides commands to copy the automation ID, edit the automation, or delete the automation.

Select the name of an automation to view **Details** about it and relevant **Events**.

### Triggers

Triggers specify the conditions under which your action should be performed. The Prefect UI includes templates for many 
common conditions, such as:

- Flow run state change (Flow Run Tags are only evaluated with `OR` criteria)
- Work pool status
- Work queue status
- Deployment status
- Metric thresholds, such as average duration, lateness, or completion percentage
- Incident declarations (available on Pro and Custom plans)
- Custom event triggers

<Note>
**Automations API**

The [automations API](https://app.prefect.cloud/api/docs#tag/Automations) enables further programmatic customization of 
trigger and action policies based on arbitrary [events](https://app.prefect.cloud/api/docs#tag/Events).
</Note>

> Importantly, you can configure the triggers not only in reaction to events, but also proactively: to fire in the absence of 

Reviewer: ```suggestion
Importantly, you can configure the triggers not only in reaction to events, but also proactively: in the absence of 
```

or
automation = Automation.read("woodchonk")

print(automation)
# name='woodchonk' description='' enabled=True trigger=EventTrigger(type='event', match=ResourceSpecification(__root__={'genus': 'Marmota', 'species': 'monax'}), match_related=ResourceSpecification(__root__={}), after=set(), expect={'animal.walked'}, for_each=set(), posture=Posture.Reactive, threshold=3, within=datetime.timedelta(seconds=10)) actions=[CancelFlowRun(type='cancel-flow-run')] actions_on_trigger=[] actions_on_resolve=[] owner_resource=None id=UUID('d641c552-775c-4dc6-a31e-541cb11137a6')
```

### Selected and inferred action targets

Some actions require you to either select the target of the action, or specify that the target of the action should be 
inferred. Selected targets are simple and useful for when you know exactly what object your action should act on. For 
example, the case of a cleanup flow you want to run or a specific notification you want to send.

Inferred targets are deduced from the trigger itself.

For example, if a trigger fires on a flow run that is stuck in a running state, and the action is to cancel an inferred 

> flow run—the flow run to cancel is inferred as the stuck run that caused the trigger to fire.

Reviewer: ```suggestion
flow run—the flow run that caused the trigger to fire.
```

```yaml
triggers:
  - enabled: true
    match:
      prefect.resource.id: my.external.resource
    expect:
      - external.resource.pinged
    parameters:
      param_1: "{{ event }}"
  - enabled: true
    match:
      prefect.resource.id: my.other.external.resource
    expect:
      - some.other.event
    parameters:
      param_1: "{{ event }}"
```

Both of the above triggers would be attached to `test-deployment` after running `prefect deploy`.

<Warning>
**Triggers passed to `prefect deploy` will override any triggers defined in `prefect.yaml`**

While you can define triggers in `prefect.yaml` for a given deployment, triggers passed to `prefect deploy` 
take precedence over those defined in `prefect.yaml`.
</Warning>

Note that deployment triggers contribute to the total number of automations in your workspace.

## Automation notifications

Notifications enable you to set up automation actions that send a message.

Automation notifications support sending notifications through any predefined block that is capable of and configured 
to send a message. That includes, for example:

- Slack message to a channel
- Microsoft Teams message to a channel

> - Email to a configured email address

Reviewer: ```suggestion
- Email to an email address
```

workflow logic. 

1. Prior to creating the automation, confirm the notification location. Create a notification block to help define where
the notification is sent. 
![List of available blocks](/3.0rc/img/guides/block-list.png)

2. Navigate to the blocks page on the UI, and click into creating an email notification block. 
![Creating a notification block in the Cloud UI](/3.0rc/img/guides/notification-block.png)

3. Go to the automations page to create your first automation.
![Automations page](/3.0rc/img/guides/automation-list.png)

4. Next, find the trigger type. In this case, use a flow completion. 
![Trigger type](/3.0rc/img/guides/automation-triggers.png)

5. Create the actions for when the trigger is hit. In this case, create a notification to showcase the 
completion. 
![Notification block in automation](/3.0rc/img/guides/notify-auto-block.png)

6. Now the automation is ready to be triggered from a flow run completion. Run the file locally and see that the 
notification is sent to your inbox after the completion. It may take a few minutes for the notification to arrive.
![Final notification](/3.0rc/img/guides/final-automation.png)

<Tip>
**No deployment created**

> You do not need to create a deployment to trigger your automation, where a state outcome of a local 

Reviewer: ```suggestion
You do not need to create a deployment to trigger your automation. In the case above, the flow run state trigger fired in response to a flow run that was executed locally.
```

> flow run helped trigger this notification block. You are not required to create a deployment to trigger a notification.

Reviewer: ```suggestion
```

```console
curl -X POST https://api.prefect.cloud/hooks/34iV2SFke3mVa6y5Y-YUoA -d "model_id=adhoc" -d "run_count=10" -d "friendly_name=test-user-input"
```
From here, you can make a webhook that is connected to pulling in parameters on the curl command. It kicks off a 
deployment that uses these pulled parameters:
![Webhook created](/3.0rc/img/guides/webhook-created.png)

Go into the event feed to automate straight from this event: 
![Webhook automate](/3.0rc/img/guides/webhook-automate.png)

This allows you to create automations that respond to these webhook events. From a few clicks in the UI, you can 
associate an external process with the Prefect events API that can trigger downstream deployments. 
![Automation custom](/3.0rc/img/guides/automation-custom.png)

## Use triggers

Explore event triggers that automate the kickoff of a deployment run.

You can create a deployment that triggers when a flow run takes longer than expected.

### Use the Marvin library

Take advantage of Prefect's [Marvin](https://www.askmarvin.ai/) library that uses an LLM to classify the data. 

> Marvin is great at embedding data science and data analysis applications within your pre-existing data engineering workflows. 

Reviewer: ```suggestion
```

> In this case, you can use [Marvin's AI functions](https://www.askmarvin.ai/docs/text/functions/) to help make the dataset more information rich. 

Reviewer: ```suggestion
You can use [Marvin's AI functions](https://www.askmarvin.ai/docs/text/functions/) to make the dataset more information rich. 
```
